Products/CMFPlone/recyclebin.py

```diff
@@ -0,0 +1,584 @@
+            ),
+            "title": item_title,
+            "type": item_type or getattr(obj, "portal_type", "Unknown"),
+            "path": original_path,
+            "parent_path": "/".join(original_container.getPhysicalPath()),
+            "deletion_date": datetime.now(),
+            "size": getattr(obj, "get_size", lambda: 0)(),
+            "object": obj,  # Store the actual object
+        }
+
+        # Check if we need to clean up old items
+        self._check_size_limits()
+
+        return item_id
+
+    def get_items(self):
+        """Return all items in recycle bin"""
+        items = []
+        for item_id, data in self.storage.items():
+            item_data = data.copy()
+            item_data["recycle_id"] = item_id
+            # Don't include the actual object in the listing
+            if "object" in item_data:
+                del item_data["object"]
+            items.append(item_data)
+
+        # Sort by deletion date (newest first)
+        return sorted(items, key=lambda x: x["deletion_date"], reverse=True)
+
+    def get_item(self, item_id):
+        """Get a specific deleted item by ID"""
+        return self.storage.get(item_id)
+
+    def restore_item(self, item_id, target_container=None):
+        """Restore item to original location or specified container"""
+        if item_id not in self.storage:
+            return None
+
+        item_data = self.storage[item_id]
+        obj = item_data["object"]
+        obj_id = item_data["id"]
+        item_type = item_data.get("type", None)
+
+        # Special handling for CommentTree (comments with replies)
+        if item_type == "CommentTree":
+            return self._restore_comment_tree(item_id, item_data, target_container)
+
+        # Special handling for Discussion Item (Comments)
+        if item_data.get("type") == "Discussion Item":
+            return self._restore_comment(item_id, item_data, target_container)
+
+        # Regular content object restoration
+        # Find the container to restore to
+        site = self._get_context()
+        if target_container is None:
+            # Try to get the original parent
+            parent_path = item_data["parent_path"]
+            try:
+                target_container = site.unrestrictedTraverse(parent_path)
+            except (KeyError, AttributeError):
+                # If original parent doesn't exist, restore to site root
+                target_container = site
+
+        # Make sure we don't overwrite existing content
+        if obj_id in target_container:
+            # Generate a unique ID by appending a timestamp
+            timestamp = datetime.now().strftime("%Y%m%d%H%M%S")
+            obj_id = f"{obj_id}-restored-{timestamp}"
+
+        # Set the new ID if it was changed
+        if obj_id != item_data["id"]:
+            obj.id = obj_id
+
+        # Add object to the target container
+        target_container._setObject(obj_id, obj)
```

Comment:
Usually we write it this way:

```suggestion
        target_container[obj_id] = obj
```

That's the same as calling _setObject because of https://github.com/zopefoundation/Zope/blob/master/src/OFS/ObjectManager.py#L747